hw/ip/otbn/dv/rig/rig/gens/bad_load_store.py

```diff
@@ -0,0 +1,299 @@
+            # very often.
+            prog_insn = self.fill_insn(self.insns[idx], model)
+            if prog_insn is None:
+                weights = self.weights.copy()
+                weights[idx] = 0
+                continue
+
+        snippet = ProgSnippet(model.pc, [prog_insn])
+        snippet.insert_into_program(program)
+
+        return (snippet, True, model)
+
+    def fill_insn(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+        '''Try to pick one of BN.XID or XW instructions
+
+        '''
+
+        # Special-case BN load/store instructions by mnemonic. These use
+        # complicated indirect addressing, so it's probably more sensible to
+        # give them special code.
+        if insn.mnemonic in ['bn.lid', 'bn.sid']:
+            return self._fill_bn_xid(insn, model)
+        if insn.mnemonic in ['lw', 'sw']:
+            return self._fill_xw(insn, model)
+
+        return None
+
+    def _fill_xw(self, insn: Insn, model: Model) -> Optional[ProgInsn]:
+
+        grd_op_type = self.lw.operands[0].op_type
+        imm_op_type = self.lw.operands[1].op_type
+
+        # Determine the offset range for XW
+        offset_rng = imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        # Max/Min Offsets for XW (-2048, 2047)
+        min_offset, max_offset = offset_rng
+
+        # Get known registers
+        known_regs = model.regs_with_known_vals('gpr')
+
+        # We have two options for this generator: Barely OOB or OOB
+        barely_oob = []
+        oob = []
+
+        for reg_idx, u_reg_val in known_regs:
+            reg_val = u_reg_val - (1 << 32) if u_reg_val >> 31 else u_reg_val
+            # We can reach to barely OOB by adding max_offset to smallest val
+            # or adding min_offset (which is negative) to biggest possible val
+            if reg_val - model.dmem_size in range(-max_offset, -min_offset):
+                barely_oob.append((reg_idx, reg_val))
+            # Negative OOB
+            if reg_val + min_offset < 0:
+                oob.append((reg_idx, reg_val))
+            # Positive OOB
+            if reg_val + max_offset > model.dmem_size:
+                oob.append((reg_idx, reg_val))
+
+        if random.random() < 0.2 and barely_oob:
+            idx, val = random.choice(barely_oob)
+            tgt_addr = model.dmem_size
+        elif oob:
+            idx, val = random.choice(oob)
+            if val + min_offset < 0:
+                tgt_addr = random.randrange(val + min_offset, -1)
+            else:
+                tgt_addr = random.randrange(model.dmem_size, val + max_offset)
+        else:
+            return None
+
+        # If u_val makes target address very much out of offset range
+        # just randomize imm_val because oob address is already guaranteed
+        # and we still want to have a valid imm_val
+        if tgt_addr - val not in range(min_offset, max_offset):
+            imm_val = random.randint(min_offset, max_offset)
```

Comment:
Can this ever happen? The code above looks like it's choosing things so that `tgt_addr` is representable?

Reply:
Let's say `val = 2^16 - 1`. In that case whatever the immediate is, we would be in OOB region. The piece is needed because we calculate `imm_val` from `tgt_addr`  and if it is too big it would throw an error while generating.

Reply:
What I mean is that the code on lines 197-205 picks a base address (represented by `(idx, val)`) and a target address (`tgt_addr`) that I *think* ensures `tgt_addr - val` is representable. So I don't think this code should ever run.